In [1]:
import pandas as pd
ratings = pd.read_csv('ratings.csv')
movies = pd.read_csv('movies.csv')
links = pd.read_csv('links.csv')

In [2]:
ratings.head(2)

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247

In [3]:
ratings.rating.describe()

count    100836.000000
mean          3.501557
std           1.042529
min           0.500000
25%           3.000000
50%           3.500000
75%           4.000000
max           5.000000
Name: rating, dtype: float64

In [4]:
ratings.movieId.head(10)

0      1
1      3
2      6
3     47
4     50
5     70
6    101
7    110
8    151
9    157
Name: movieId, dtype: int64

In [5]:
ratings.shape

(100836, 4)

In [6]:
movies.head(2)

movieId             title                                       genres
0        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy
1        2    Jumanji (1995)                   Adventure|Children|Fantasy

In [7]:
movies.movieId.head(10)

0     1
1     2
2     3
3     4
4     5
5     6
6     7
7     8
8     9
9    10
Name: movieId, dtype: int64

In [8]:
movies.shape

(9742, 3)

In [9]:
links.head(2)

movieId  imdbId  tmdbId
0        1  114709   862.0
1        2  113497  8844.0

In [10]:
links.movieId.head(10)

0     1
1     2
2     3
3     4
4     5
5     6
6     7
7     8
8     9
9    10
Name: movieId, dtype: int64

In [11]:
links.shape

(9742, 3)

In [12]:
movies_links=movies.merge(links, left_on='movieId',right_on='movieId')
movies_links.sort_values(['movieId'], ascending=True)

movieId                                      title  \
0           1                           Toy Story (1995)   
1           2                             Jumanji (1995)   
2           3                    Grumpier Old Men (1995)   
3           4                   Waiting to Exhale (1995)   
4           5         Father of the Bride Part II (1995)   
...       ...                                        ...   
9737   193581  Black Butler: Book of the Atlantic (2017)   
9738   193583               No Game No Life: Zero (2017)   
9739   193585                               Flint (2017)   
9740   193587        Bungo Stray Dogs: Dead Apple (2018)   
9741   193609        Andrew Dice Clay: Dice Rules (1991)   

                                           genres   imdbId    tmdbId  
0     Adventure|Animation|Children|Comedy|Fantasy   114709     862.0  
1                      Adventure|Children|Fantasy   113497    8844.0  
2                                  Comedy|Romance   113228   15602.0  
3                            Comedy|Drama|Romance   114885   31357.0  
4                                          Comedy   113041   11862.0  
...                                           ...      ...       ...  
9737              Action|Animation|Comedy|Fantasy  5476944  432131.0  
9738                     Animation|Comedy|Fantasy  5914996  445030.0  
9739                                        Drama  6397426  479308.0  
9740                             Action|Animation  8391976  483455.0  
9741                                       Comedy   101726   37891.0  

[9742 rows x 5 columns]

In [13]:
genres = set()
for genre_string in movies_links['genres']:
   genre_list = genre_string.split('|')
   for genre in genre_list:
       genres.add(genre)
genres = list(genres)

In [14]:
for g in genres:
    movies_links[g]=movies_links['genres'].apply(lambda x: g in x)

In [15]:
genres

['Comedy',
 'Musical',
 'Children',
 'Animation',
 'Horror',
 'Film-Noir',
 'Western',
 'Romance',
 'Thriller',
 'War',
 '(no genres listed)',
 'Adventure',
 'Fantasy',
 'Documentary',
 'Action',
 'Sci-Fi',
 'Drama',
 'Mystery',
 'IMAX',
 'Crime']

In [16]:
len(genres)

20

In [17]:
movies_links_combined = movies_links.filter(['movieId','title','genres','imdbId','tmdbId','Drama','Mystery','Western','Documentary','IMAX','Sci-Fi',
'Children','Romance','Musical','Comedy','Thriller','Fantasy','Action','Animation','War',
'Adventure','(no genres listed)','Horror','Film-Noir','Crime'])

In [18]:
#This will convert the True and False values for the genres into 1's and 0's.
movies_links_combined[['Drama','Mystery','Western','Documentary','IMAX','Sci-Fi',
'Children','Romance','Musical','Comedy','Thriller','Fantasy','Action','Animation','War',
'Adventure','(no genres listed)','Horror','Film-Noir','Crime']] *= 1

In [19]:
movies_links_combined.head(2)

movieId             title                                       genres  \
0        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
1        2    Jumanji (1995)                   Adventure|Children|Fantasy   

   imdbId  tmdbId  Drama  Mystery  Western  Documentary  IMAX  ...  Thriller  \
0  114709   862.0      0        0        0            0     0  ...         0   
1  113497  8844.0      0        0        0            0     0  ...         0   

   Fantasy  Action  Animation  War  Adventure  (no genres listed)  Horror  \
0        1       0          1    0          1                   0       0   
1        1       0          0    0          1                   0       0   

   Film-Noir  Crime  
0          0      0  
1          0      0  

[2 rows x 25 columns]

In [20]:
#The genres column is no longer necessary with the new, binary genre information.
movies_links_combined.drop('genres',axis=1,inplace=True)
movies_links_combined.drop('(no genres listed)',axis=1,inplace=True)
movies_links_combined.head(2)

movieId             title  imdbId  tmdbId  Drama  Mystery  Western  \
0        1  Toy Story (1995)  114709   862.0      0        0        0   
1        2    Jumanji (1995)  113497  8844.0      0        0        0   

   Documentary  IMAX  Sci-Fi  ...  Comedy  Thriller  Fantasy  Action  \
0            0     0       0  ...       1         0        1       0   
1            0     0       0  ...       0         0        1       0   

   Animation  War  Adventure  Horror  Film-Noir  Crime  
0          1    0          1       0          0      0  
1          0    0          1       0          0      0  

[2 rows x 23 columns]

In [21]:
df_combined=movies_links_combined.merge(ratings, left_on='movieId',right_on='movieId')
df_combined.sort_values(['movieId'], ascending=True)

movieId                                      title   imdbId    tmdbId  \
0             1                           Toy Story (1995)   114709     862.0   
137           1                           Toy Story (1995)   114709     862.0   
138           1                           Toy Story (1995)   114709     862.0   
139           1                           Toy Story (1995)   114709     862.0   
140           1                           Toy Story (1995)   114709     862.0   
...         ...                                        ...      ...       ...   
100831   193581  Black Butler: Book of the Atlantic (2017)  5476944  432131.0   
100832   193583               No Game No Life: Zero (2017)  5914996  445030.0   
100833   193585                               Flint (2017)  6397426  479308.0   
100834   193587        Bungo Stray Dogs: Dead Apple (2018)  8391976  483455.0   
100835   193609        Andrew Dice Clay: Dice Rules (1991)   101726   37891.0   

        Drama  Mystery  Western  Documentary  IMAX  Sci-Fi  ...  Action  \
0           0        0        0            0     0       0  ...       0   
137         0        0        0            0     0       0  ...       0   
138         0        0        0            0     0       0  ...       0   
139         0        0        0            0     0       0  ...       0   
140         0        0        0            0     0       0  ...       0   
...       ...      ...      ...          ...   ...     ...  ...     ...   
100831      0        0        0            0     0       0  ...       1   
100832      0        0        0            0     0       0  ...       0   
100833      1        0        0            0     0       0  ...       0   
100834      0        0        0            0     0       0  ...       1   
100835      0        0        0            0     0       0  ...       0   

        Animation  War  Adventure  Horror  Film-Noir  Crime  userId  rating  \
0               1    0          1       0          0      0       1     4.0   
137             1    0          1       0          0      0     382     4.5   
138             1    0          1       0          0      0     385     4.0   
139             1    0          1       0          0      0     389     5.0   
140             1    0          1       0          0      0     391     3.0   
...           ...  ...        ...     ...        ...    ...     ...     ...   
100831          1    0          0       0          0      0     184     4.0   
100832          1    0          0       0          0      0     184     3.5   
100833          0    0          0       0          0      0     184     3.5   
100834          1    0          0       0          0      0     184     3.5   
100835          0    0          0       0          0      0     331     4.0   

         timestamp  
0        964982703  
137     1515162628  
138      834691642  
139      857934174  
140     1032388077  
...            ...  
100831  1537109082  
100832  1537109545  
100833  1537109805  
100834  1537110021  
100835  1537157606  

[100836 rows x 26 columns]

In [22]:
df_combined.shape

(100836, 26)

In [23]:
df_combined.columns

Index(['movieId', 'title', 'imdbId', 'tmdbId', 'Drama', 'Mystery', 'Western',
       'Documentary', 'IMAX', 'Sci-Fi', 'Children', 'Romance', 'Musical',
       'Comedy', 'Thriller', 'Fantasy', 'Action', 'Animation', 'War',
       'Adventure', 'Horror', 'Film-Noir', 'Crime', 'userId', 'rating',
       'timestamp'],
      dtype='object')

In [24]:
#Rearranging the columnns.
df_combined = df_combined[['movieId','userId','rating','timestamp','title','imdbId','tmdbId',
'Drama','Mystery','Western','Documentary','IMAX','Sci-Fi','Children','Romance','Musical',
'Comedy','Thriller','Fantasy','Action','Animation','War','Adventure','Horror','Film-Noir', 'Crime']]

In [25]:
df_combined.head(2)

movieId  userId  rating  timestamp             title  imdbId  tmdbId  \
0        1       1     4.0  964982703  Toy Story (1995)  114709   862.0   
1        1       5     4.0  847434962  Toy Story (1995)  114709   862.0   

   Drama  Mystery  Western  ...  Comedy  Thriller  Fantasy  Action  Animation  \
0      0        0        0  ...       1         0        1       0          1   
1      0        0        0  ...       1         0        1       0          1   

   War  Adventure  Horror  Film-Noir  Crime  
0    0          1       0          0      0  
1    0          1       0          0      0  

[2 rows x 26 columns]

In [26]:
#Do I be want the first column, or should movieId be my index? Yish: This does not matter but probably do not
#want to change this because multiple movieIds (they appear more than once).
df_combined.set_index(['movieId'])
df_combined.head(2)
#This did not seem to change anything.

movieId  userId  rating  timestamp             title  imdbId  tmdbId  \
0        1       1     4.0  964982703  Toy Story (1995)  114709   862.0   
1        1       5     4.0  847434962  Toy Story (1995)  114709   862.0   

   Drama  Mystery  Western  ...  Comedy  Thriller  Fantasy  Action  Animation  \
0      0        0        0  ...       1         0        1       0          1   
1      0        0        0  ...       1         0        1       0          1   

   War  Adventure  Horror  Film-Noir  Crime  
0    0          1       0          0      0  
1    0          1       0          0      0  

[2 rows x 26 columns]

In [27]:
#Dropping unnecessary columns.
df_combined = df_combined.drop(columns='timestamp')
#What are imdbId and tmdbId? Do we need these if we also have movieId? Potentially redundant.
df_combined.shape

(100836, 25)

In [28]:
from surprise import Reader, Dataset
reader = Reader()
data = Dataset.load_from_df(df_combined[['movieId','userId','rating']], reader)

In [29]:
dataset = data.build_full_trainset()
print('Number of users: ', dataset.n_users,)
print('Number of items: ', dataset.n_items)

Number of users:  9724
Number of items:  610


In [30]:
from surprise.model_selection import cross_validate
from surprise.prediction_algorithms import SVD
from surprise.prediction_algorithms import KNNWithMeans, KNNBasic, KNNBaseline
from surprise.model_selection import GridSearchCV
import numpy as np

In [31]:
#Performing a gridsearch with SVD.
params = {'n_factors': [10, 25, 50],
         'reg_all': [0.01, 0.025, 0.05]}
g_s_svd = GridSearchCV(SVD,param_grid=params,n_jobs=-1)
g_s_svd.fit(data)

In [32]:
print(g_s_svd.best_score)
print(g_s_svd.best_params)

{'rmse': 0.8740526758506532, 'mae': 0.6723665001076992}
{'rmse': {'n_factors': 50, 'reg_all': 0.05}, 'mae': {'n_factors': 25, 'reg_all': 0.025}}


Can re-run SVD with each (25 factors vs. 50) and compare.

In [33]:
#Cross-validating with KNNBasic.
knn_basic = KNNBasic(sim_options={'name':'pearson', 'user_based':True})
cv_knn_basic = cross_validate(knn_basic, data, n_jobs=-1)

In [34]:
#Print out the average RMSE score for the test set.
for i in cv_knn_basic.items():
    print(i)
print('\n')
print(np.mean(cv_knn_basic['test_rmse']))

('test_rmse', array([0.97241185, 0.96602046, 0.96296606, 0.97327269, 0.97291619]))
('test_mae', array([0.75497333, 0.7521774 , 0.74489811, 0.75650336, 0.75782592]))
('fit_time', (312.47456216812134, 322.14636397361755, 319.95260405540466, 322.2047040462494, 66.20753192901611))
('test_time', (55.97702717781067, 55.83063292503357, 56.47214412689209, 54.756025075912476, 12.932823181152344))


0.9695174494938421


In [35]:
#Cross-validating with KNNBaseline.
knn_baseline = KNNBaseline(sim_options={'name':'pearson', 'user_based':True})
cv_knn_baseline = cross_validate(knn_baseline,data)

Estimating biases using als...
Computing the pearson similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson similarity matrix...
Done computing similarity matrix.


In [36]:
#Printing the average RMSE for the test set.
for i in cv_knn_baseline.items():
    print(i)

np.mean(cv_knn_baseline['test_rmse'])

('test_rmse', array([0.88482918, 0.88489929, 0.88008807, 0.88950738, 0.88458049]))
('test_mae', array([0.67959921, 0.68183005, 0.67823298, 0.68334574, 0.67994086]))
('fit_time', (75.29813385009766, 55.733351945877075, 44.77215909957886, 36.863041162490845, 38.576772928237915))
('test_time', (25.35781979560852, 16.42236614227295, 12.897680282592773, 13.209712028503418, 13.4442880153656))


0.884780883318998

On average, we would expect the estimates from our model to be off by about .88 points per review.